File: year4/rFEM/geometry/edge.rb
require File.expand_path(File.dirname(__FILE__)) + '/point.rb'

class Edge
  attr :origin, true
  attr :destination, true

  def initialize(*args)
    case args.size
    when 0 then
        @origin = Point.new(0, 0)
      @destination = Point.new(0, 0)
    when 1 then
        raise ArgumentError.new('Argument is not an Edge') unless args[0].instance_of? Edge
      @origin = Point.new(args[0].origin.x, args[0].origin.y)
      @destination = Point.new(args[0].destination.x, args[0].destination.y)
    when 2 then
        raise ArgumentError.new('Arguments are not Points') unless args[0].instance_of?(Point) and args[1].instance_of?(Point)
      @origin = Point.new(args[0].x, args[0].y)
      @destination = Point.new(args[1].x, args[1].y)
    end
  end

  # rotate on 90* clockwise
  def rotate!
    m = (@origin + @destination)*0.5
    v = @destination - @origin

    n = Point.new(v.y, -v.x)

    @origin = m - n*0.5
    @destination = m + n*0.5
  end

  def ==(e)
    (@origin == e.origin) and (@destination == e.destination)
  end

  def length
    Math.sqrt((@origin.x - @destination.x)**2 +
              (@destination.y - @origin.y)**2)
  end

  def flip!
    rotate!
    rotate!
  end

  def intersect (edge)
    a = @origin
    b = @destination

    c = edge.origin
    d = edge.destination

    n = Point.new((d - c).y, (c - d).x)
    denominator = n*(b - a)

#    if denominator.zero?
#      aclass = @origin.classify(edge.origin, edge.destination)
#      if (aclass == :left) or (aclass == :right)
#        return :parallel
#      else
#        return :collinear
#      end
    #    end
    if denominator.zero?
      raise 'denominator is zero!'
    end

    num = n*(a - c)
    -num/denominator
    
  end

  def Edge.length(p1, p2)
    Math.sqrt((p1.x - p2.x)**2 + (p1.y - p2.y)**2)
  end

  def Edge.angle(edge1, edge2)
    # just use the cosine theorem
    len1 = edge1.length
    len2 = edge2.length

    len3 = Edge.length(edge1.origin, edge2.destination)

    value = (len1*len1 + len2*len2 - len3*len3)/(2.0*len1*len2)

    value = -1.0 if value < -1.0
    value = 1.0 if value > 1.0

    Math.acos(value).abs
  end

  def Edge.angle_points(point1, point2, point3)
    len1 = Edge.length point1, point2
    len2 = Edge.length point2, point3

    len3 = Edge.length point1, point3

    value = (len1*len1 + len2*len2 - len3*len3)/(2.0*len1*len2)

    value = -1.0 if value < -1.0
    value = 1.0 if value > 1.0

    Math.acos(value).abs
  end

  def Edge.angle_points_orient(point_next, point, point_prev)
    angle = Edge.angle_points point_prev, point, point_next
    
    if point_next.classify(point_prev, point) == :left
      angle = Math::PI*2.0 - angle
    end
    angle
  end

  def Edge.split(point1, point2, diameter)
    points_array = []

    length = Edge.length(point1, point2)
      
    points_count = (length/diameter).ceil
    step = length / points_count
    
    angle = Math.atan((point2.y - point1.y).abs/(point2.x - point1.x).abs)
    step_x = step * Math.cos(angle)
    step_y = step * Math.sin(angle)
    
    sign_x = 1.0
    sign_x = -1.0 if (point2.x - point1.x) < 0
    
    sign_y = 1.0
    sign_y = -1.0 if (point2.y - point1.y) < 0
    
    points_count.times do |i|
      points_array << Point.new(point1.x + i*step_x*sign_x,
                      point1.y + i*step_y*sign_y)
    end

    points_array
  end
  
end

File: year4/pyFEM/geometry/edge.py
import math

from pyFEM.geometry import Point
from pyFEM import symbols as sym


class Edge(object):
    def __init__(self, *args):
        if len(args) == 0:
            self.origin = Point(0, 0)
            self.destination = Point(0, 0)
        elif len(args) == 1:
            e = args[0]
            self.origin = Point(e.origin.x, e.origin.y)
            self.destination = Point(e.destination.x, e.destination.y)
        elif len(args) == 2:
            self.origin, self.destination = [Point(p.x, p.y) for p in args]
        else:
            raise ValueError("Unexpected number of arguments")

    def __eq__(self, other):
        return (self.origin == other.origin) and (self.destination == other.destination)

    def length(self):
        return (destination - origin).length()

    def rotate(self):
        """Rotate self 90 degrees CW around the center"""
        m = (self.origin + self.destination) * 0.5
        v = self.destination - self.origin
        n = Point(v.y, -v.x)
        self.origin = m - n * 0.5
        self.destination = m + n * 0.5

    def flip(self):
        self.rotate()
        self.rotate()

    def intersect(self, other):
        raise NotImplementedError("Edge.intersect stub")

    @staticmethod
    def angle(edge1, edge2):
        raise NotImplementedError("Edge.angle stub")

    @staticmethod
    def angle_points(point1, point2, point3):
        """Return the absolute value of the angle in radians
        between [point1, point2] and [point2, point3]"""
        # use law of cosines
        a = (point2 - point1).length()
        b = (point3 - point2).length()
        c = (point3 - point1).length()

        value = ((a ** 2) + (b ** 2) - (c ** 2)) / (2.0 * a * b)
        if abs(value) > 1.0:
            value = 1.0 * (value / abs(value))
        return abs(math.acos(value))

    @staticmethod
    def angle_points_orient(point_next, point, point_prev):
        angle = Edge.angle_points(point_prev, point, point_next)
        if point_next.classify(point_prev, point) == sym.left:
            angle = (math.pi * 2.0) - angle
        return angle
    
    @staticmethod
    def split(point1, point2, diameter):
        # TODO: implement `split' in terms of point count instead of diameter
        l = (point2 - point1).length()
        n = int(math.ceil(l / diameter))
        step_v = (point2 - point1) * (1.0 / n)
        return [point1 + (step_v * i) for i in range(n)]

    @staticmethod
    def subdivide(point1, point2, point_count):
        step_v = (point2 - point1) * (1.0 / (point_count + 1))
        return [point1 + (step_v * i) for i in range(1, point_count + 1)]

    @staticmethod
    def subdivide_with_bounds(point1, point2, point_count):
        res = [point1]
        res.extend(Edge.subdivide(point1, point2, point_count))
        res.append(point2)
        return res
        
  
